Cody: Remove server endpoint from user settings

deepak2431 opened the pull request:
This PR is for #51869. The following changes are included in this PR:
1. Removed server endpoint from the user settings
2. Used the login settings retrieved in the webview to save in the secret storage and retrieve from there for all-purpose
3. Updated unit tests
4. Updated integration tests

## Test plan
- All integration tests have been passed
- All unit tests have passed
- Tested the actual behavior by login/logout the Cody and it works completely fine with the current implementation in this PR.

https://github.com/sourcegraph/sourcegraph/assets/44617923/31166be2-ebd7-4229-96d9-634f8c189fa6

deepak2431 commented:
@philipp-spiess @abeatrix Requesting review for this PR :)

eseliger commented on client/cody-shared/src/telemetry/EventLogger.ts:
```diff
@@ -1,15 +1,15 @@
-import * as vscode from 'vscode'
+// import * as vscode from 'vscode'
```
why commented out vs removing?

deepak2431 replied:
Thanks for the review @eseliger. I have commented this out because there’s a small TODO left to be done here. If you see the PR description I have mentioned it that what would be the best solution to send the logs as we now don’t have the Cody server endpoint in the vs code settings.

eseliger replied:
hah, sorry for missing that 😞 

Initializing the logger with the endpoint sounds reasonable to me, but let me also ping @abeatrix and @philipp-spiess who are deeper into the extension :)

deepak2431 replied:
no problem @eseliger. For me too passing the server directly in logger function seems a better solution. Though with wait for the views from @abeatrix  and @philipp-spiess

philipp-spiess replied:
+1 on passing the URL in the constructor. This way it would also be easier to test it 🙂

eseliger commented on client/cody/src/configuration.test.ts:
```diff
@@ -2,32 +2,32 @@ import type * as vscode from 'vscode'
 
 import { getConfiguration } from './configuration'
 
+// Fix this tests
```
is this leftover instructions to cody or are these notes to yourself? same in 24

deepak2431 replied:
This was just notes for me to remember, I will remove this.

thenamankumar replied:
Try if Cody can solve this for you.

abeatrix commented on client/cody/src/configuration.test.ts:
```diff
@@ -40,6 +40,8 @@ describe('getConfiguration', () => {
                         return true
                     case 'cody.experimental.inline':
                         return true
+                    case 'cody.experimental.connectToApp':
```
this has been removed

abeatrix commented on client/cody/src/main.ts:
```diff
@@ -161,6 +167,9 @@ const register = async (
             }
         }),
+        vscode.commands.registerCommand('cody.set-serverEndpoint', async (serverEndpoint: string) => {
+            await secretStorage.store(CODY_SERVER_ENDPOINT, serverEndpoint)
```
We should store the endpoints using the storage API instead of the secret storage API, especially with the Secret Storage corruption issue we have been encountering.

abeatrix reviewed (changes requested):
Thanks for working on this issue @deepak2431 ! 

As mentioned in one of my comments, I think we should store the endpoints using the storage API instead of the secret storage API due to the Secret Storage corruption issue we have been encountering. That also allows us to support multiple endpoints easier as described in the [issue](https://github.com/sourcegraph/sourcegraph/issues/51869), what do you think? 

We should also automatically migrate the current server endpoint setting from the config to the storage for the user so that they won't be logged out because of the new change 😄

deepak2431 commented:
> Thanks for working on this issue @deepak2431 !
> 
> As mentioned in one of my comments, I think we should store the endpoints using the storage API instead of the secret storage API due to the Secret Storage corruption issue we have been encountering. That also allows us to support multiple endpoints easier as described in the [issue](https://github.com/sourcegraph/sourcegraph/issues/51869), what do you think?
> 
> We should also automatically migrate the current server endpoint setting from the config to the storage for the user so that they won't be logged out because of the new change 😄



> Thanks for working on this issue @deepak2431 !
> 
> As mentioned in one of my comments, I think we should store the endpoints using the storage API instead of the secret storage API due to the Secret Storage corruption issue we have been encountering. That also allows us to support multiple endpoints easier as described in the [issue](https://github.com/sourcegraph/sourcegraph/issues/51869), what do you think?
> 
> We should also automatically migrate the current server endpoint setting from the config to the storage for the user so that they won't be logged out because of the new change 😄

Thanks for the review @abeatrix. That sounds good to me we can store the server endpoint in the storage API. Let me work on updating the telemetry accordingly. 

Also, regarding the migration of the users I don’t think so that they will be logged out because it’s in sync with the login settings when the access token are loaded. And even during testing of it I was not logged out but yeah I will check the behaviour again and see if that’s the case.

abeatrix commented:
> Also, regarding the migration of the users I don’t think so that they will be logged out because it’s in sync with the login settings when the access token are loaded. And even during testing of it I was not logged out but yeah I will check the behaviour again and see if that’s the case.

Oh I mean when a currently logged in users installs the new version that looks for the endpoint that is stored in the secret storage when their current one is stored in their config, wouldn't they be logged out because the endpoint returned from the storage will be empty? 🤔

deepak2431 commented:
> > Also, regarding the migration of the users I don’t think so that they will be logged out because it’s in sync with the login settings when the access token are loaded. And even during testing of it I was not logged out but yeah I will check the behaviour again and see if that’s the case.
> 
> Oh I mean when a currently logged in users installs the new version that looks for the endpoint that is stored in the secret storage when their current one is stored in their config, wouldn't they be logged out because the endpoint returned from the storage will be empty? 🤔

Agree with you, they will be logged out here. But, just a thought as after the new update if we add a migration to read and update the secret if it's available in Cody config that will also return null possibly as it's removed from the settings. Is there any way around?

abeatrix commented:
> > > Also, regarding the migration of the users I don’t think so that they will be logged out because it’s in sync with the login settings when the access token are loaded. And even during testing of it I was not logged out but yeah I will check the behaviour again and see if that’s the case.
> 
> > 
> 
> > Oh I mean when a currently logged in users installs the new version that looks for the endpoint that is stored in the secret storage when their current one is stored in their config, wouldn't they be logged out because the endpoint returned from the storage will be empty? 🤔
> 
> 
> 
> Agree with you, they will be logged out here. But, just a thought as after the new update if we add a migration to read and update the secret if it's available in Cody config that will also return null possibly as it's removed from the settings. Is there any way around?

Instead of removing it from setting, I'd add a "deprecationMessage" to it instead so that users know that it's deprecated, and we can still read it and move it to the storage for them:
> If you set deprecationMessage, or markdownDeprecationMessage, the setting will get a warning underline with your specified message. It won't show up in the settings UI unless it is configured by the user.
[Source](https://code.visualstudio.com/api/references/contribution-points)

deepak2431 commented:
@abeatrix Sounds like a great workaround for this. Let me implement it that way, Thanks for your suggestions.

abeatrix commented:
> @abeatrix Sounds like a great workaround for this. Let me implement it that way, Thanks for your suggestions.

@deepak2431 No problem! Thank YOU for all your help on Cody, we really appreciate it!
Here is how we do the config to storage migration in the Sourcegraph VS Code extension for your reference: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/vscode/src/settings/accessTokenSetting.ts?L9-20
If you run into any issues or have any questions for me, feel free to ping me here or on Discord 😄

deepak2431 commented:
> > @abeatrix Sounds like a great workaround for this. Let me implement it that way, Thanks for your suggestions.
> 
> @deepak2431 No problem! Thank YOU for all your help on Cody, we really appreciate it! Here is how we do the config to storage migration in the Sourcegraph VS Code extension for your reference: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/vscode/src/settings/accessTokenSetting.ts?L9-20 If you run into any issues or have any questions for me, feel free to ping me here or on Discord 😄

@abeatrix Thanks for sharing the reference, it would be of great help. For sure will ping you if I need any help or have any questions!

deepak2431 commented:
@abeatrix @eseliger Made all the requested changes. Do have a look :)

deepak2431 commented:
@abeatrix Can you have a final look at this PR please?

abeatrix commented:
@deepak2431 I just took a look again and it looks like my [previous feedback ](https://github.com/sourcegraph/sourcegraph/pull/51910/files#diff-18be37aa45c10e3c2053fa1d249c725c051bf91c20177067ad73209092af00a9)has not been addressed yet?

It'd be very helpful if you can also create a test to show how the https://github.com/sourcegraph/sourcegraph/issues/51869 is fixed by the PR, thank you 🙇

deepak2431 commented:
@abeatrix Added the test plan in the PR description. As this PR just fixes the first part of the issue.

abeatrix reviewed (changes requested):
Left an inline comment regarding storing the endpoint in local storage instead of secret storage.

abeatrix commented on client/cody/src/configuration.ts:
```diff
@@ -66,5 +68,21 @@ export async function updateConfiguration(configKey: string, configValue: string
+    const storageEndpoint = await secretStorage.get(CODY_SERVER_ENDPOINT)
+    const oldConfigEndpoint = vscode.workspace.getConfiguration().get<string>('cody.serverEndpoint') || ''
+    if (!storageEndpoint && oldConfigEndpoint.length > 0) {
+        await secretStorage.store(CODY_SERVER_ENDPOINT, oldConfigEndpoint)
```
(if not, we will need to migrate it from the secret storage to the local storage later which wpi;d create additional works)

deepak2431 replied:
Agreed with you here. Just checked the documentation and it makes sense to me. 
I have also used secret storage at other places in the implementation. Shall I migrate those too using the local storage for the server endpoint?

abeatrix replied:
> I have also used secret storage at other places in the implementation. Shall I migrate those too using the local storage for the server endpoint?

@deepak2431 Yes, we will need to update those as well. It seems like there are still a lot of changes that need to be done from here, shall we put this on hold until we have a new design for supporting multiple endpoint ready before we can pick this up? or do you want to continue with this PR? ( i just don't want all your hard works go to waste later when we have the design and it requires another make over 😅 )

deepak2431 replied:
As you mentioned, it would be OK to put this on hold if this bug is not a big issue. As we will support the multiple endpoints soon, I think I can take care of it once the designs are finalised.

Even I don't want my work to go waste later!!!

deepak2431 commented:
> > 
> 
> The
> 
> > > @deepak2431 I just took a look again and it looks like my [previous feedback ](https://github.com/sourcegraph/sourcegraph/pull/51910/files#diff-18be37aa45c10e3c2053fa1d249c725c051bf91c20177067ad73209092af00a9)has not been addressed yet?
> > > It'd be very helpful if you can also create a test to show how the #51869 is fixed by the PR, thank you 🙇
> > 
> > 
> > @abeatrix I have addressed all of your comments in this [commit ](https://github.com/sourcegraph/sourcegraph/pull/51910/commits/d9c154645233decefd26f2aaf3d134acba42a76a). But yeah let me create a test for this and add here.
> 
> Sorry It looks like I didn't submit my review so my inline comment didn't get published 😅 I understand this is a big task and has a lot of moving pieces, so thank you for your patient with us on the review 🙇

 No problem. It's fine. Regarding the PR review, Phillip keeps me in the loop so it's completely understandable :)

sqs commented:
<kbd>*semi-automated message*</kbd> Cody's source code has moved to a new repository, https://github.com/sourcegraph/cody. If you still intend to merge this change, please reopen your PR on that new repository.

The following are some of the path mappings from the old repository (https://github.com/sourcegraph/sourcegraph) to the new repository (https://github.com/sourcegraph/cody):

- `client/cody/ -> vscode/`
- `client/cody-shared/ -> lib/shared/`
- `client/cody-ui/ -> lib/ui/`

Thank you for the PR, and sorry for the inconvenience!
